This patch fixes black screen on Apple Mac after Shell start. These devices require console to be switched to text mode.
The caveat is that it is not the case on other devices, where switching to text mode may break text output completely,
or lead to lower screen resolution. For this reason FirmwareVendor checks are performed prior to changing modes.

diff --git a/MdePkg/Library/UefiApplicationEntryPoint/ApplicationEntryPoint.c b/MdePkg/Library/UefiApplicationEntryPoint/ApplicationEntryPoint.c
index dc05b23f..4764433f 100644
--- a/MdePkg/Library/UefiApplicationEntryPoint/ApplicationEntryPoint.c
+++ b/MdePkg/Library/UefiApplicationEntryPoint/ApplicationEntryPoint.c
@@ -18,6 +18,33 @@ WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
 #include <Library/DebugLib.h>
 #include <Library/UefiBootServicesTableLib.h>
 
+STATIC
+VOID
+MacSwitchToTextMode (
+  EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_STATUS  Status;
+  EFI_STATUS (**EFIAPI ConsoleControl)(VOID *, UINT32);
+  #define EFI_CONSOLE_CONTROL_PROTOCOL_GUID \
+    { 0xf42f7782, 0x12e, 0x4c12, {0x99, 0x56, 0x49, 0xf9, 0x43, 0x4, 0xf7, 0x21} }
+  STATIC EFI_GUID  mConsoleControlProtocolGuid = EFI_CONSOLE_CONTROL_PROTOCOL_GUID;
+  //
+  // On Apple firmwares we need to use legacy console control protocol to switch to text mode.
+  // Otherwise black screen will be shown.
+  //
+  if (!StrCmp(SystemTable->FirmwareVendor, L"Apple")) {
+    Status = SystemTable->BootServices->HandleProtocol (
+      SystemTable->ConsoleOutHandle,
+      &mConsoleControlProtocolGuid,
+      (VOID **) &ConsoleControl
+      );
+    if (!EFI_ERROR (Status)) {
+      // ConsoleControl->SetMode (ConsoleControl, EfiConsoleControlScreenText);
+      ConsoleControl[1](ConsoleControl, 0);
+    }
+  }
+}
 
 /**
   Entry point to UEFI Application.
@@ -45,6 +72,8 @@ _ModuleEntryPoint (
 {
   EFI_STATUS                 Status;
 
+  MacSwitchToTextMode (SystemTable);
+
   if (_gUefiDriverRevision != 0) {
     //
     // Make sure that the EFI/UEFI spec revision of the platform is >= EFI/UEFI spec revision of the application.
